### PR TITLE
Refactors NODROP flag into TRAIT_NODROP

### PR DESCRIPTION
**Pull request by coiax**
:cl: coiax
fix: Anti-drop implants can no longer be used to drop objects that they
were not responsible for sticking to a person's hand.
fix: Backfiring with a Barnyard spellbook will now play a spooky horse sound.
refactor: Refactors the way that "NODROP" items work to a new system,
there should be no change in functionality.
/:cl:

Various items in the codebase were doing weird hoop jumps in order to
preserve the nodrop flag's state when it also wanted to change it, so I
moved it to a trait system.

I may have gone overboard with the type of unique trait sources, but
those can be changed later. My long term plan is make a general "CURSED"
nodrop origin, which means you can unlock cursed items by being hit with
a bolt of door opening or something. But that's for another PR, this has
no functionality changes, apart from some slightly modified descriptions
on cursed masks.

- Removed a bunch of redundant voice changing code for all the voice
changing animal masks, used two new clothing flags for this purpose.
- Also refactored a bit the animal masks, making new cursed subtypes that play
the sound when created.

### CODE REVIEW

**Comment by optimumtact** [comment]
where is the so called weird hoop jumps that you're supposedly removing

**Comment by optimumtact** [comment]
ok

File: code/game/objects/items/granters.dm [hunk 1]
```diff
@@ -282,10 +282,7 @@
 /obj/item/book/granter/spell/barnyard/recoil(mob/living/carbon/user)
 	if(ishuman(user))
 		to_chat(user,"<font size='15' color='red'><b>HORSIE HAS RISEN</b></font>")
-		var/obj/item/clothing/mask/horsehead/magichead = new /obj/item/clothing/mask/horsehead
-		magichead.item_flags |= NODROP		//curses!
-		magichead.flags_inv &= ~HIDEFACE //so you can still see their face
-		magichead.voicechange = TRUE	//NEEEEIIGHH
+		var/obj/item/clothing/magichead = new /obj/item/clothing/mask/horsehead/cursed(get_turf(user))
```
Comment (ShizCalev) [inline code/game/objects/items/granters.dm]: Droploc

**Comment by k-k-karma** [comment]
this is dumb, nodrop is simple and works well enough. now i have to add a new define to a file if i want a file to not nodrop and shit

**Comment by xxalpha** [comment]
Can't you just use initial(flags) to reset an item flags instead of doing all this?

**Comment by Cyberboss** [comment]
Parroting what I said a while back earlier:

`NODROP` as a flag breaks down when more than one source tries to apply it to the same object (this includes the object itself). Converting it to a trait prevents that due to the source tracking. I'm for this solely for that reason

**Comment by coiax** [comment]
@Cyberboss You going to merge this then?

**Pull request merged**